### PR TITLE
Delegation contract v0.3.0

### PR DESCRIPTION
**Pull request by iulianpascalau**
- delegation contract v0.3.0

### CODE REVIEW

File: genesis/process/intermediate/standardDelegationProcessor.go [hunk 1]
```diff
@@ -171,8 +166,8 @@ func (sdp *standardDelegationProcessor) setDelegationStartParameters(smartContra
 	return nil
 }
 
-func (sdp *standardDelegationProcessor) executeSetNumNodes(numNodes int, sc genesis.InitialSmartContractHandler) error {
-	setNumNodesTxData := fmt.Sprintf("%s@%x", setNumNodesFunction, numNodes)
+func (sdp *standardDelegationProcessor) executeSetServiceFee(serviceFee int, sc genesis.InitialSmartContractHandler) error {
```
Comment (sasurobert) [inline genesis/process/intermediate/standardDelegationProcessor.go:169]: wasn't this supposed to be set in constructor ? why is this needed separately.

File: genesis/process/intermediate/standardDelegationProcessor.go [hunk 2]
```diff
@@ -401,54 +407,96 @@ func (sdp *standardDelegationProcessor) verifyStakedValue(sc genesis.InitialSmar
...
-	vmOutputBlsKeys, err := sdp.queryService.ExecuteQuery(scQueryBlsKeys)
+	vmOutputStakeValue, err := sdp.queryService.ExecuteQuery(scQueryStakeValue)
 	if err != nil {
 		return err
 	}
+	if len(vmOutputStakeValue.ReturnData) != 1 {
+		return fmt.Errorf("%w return data should have contained one element", genesis.ErrWhileVerifyingDelegation)
+	}
+	scStakedValue := big.NewInt(0).SetBytes(vmOutputStakeValue.ReturnData[0])
+
```
Comment (sasurobert) [inline genesis/process/intermediate/standardDelegationProcessor.go:439]: delete empty line.

File: genesis/process/intermediate/standardDelegationProcessor.go [hunk 3]
```diff
@@ -401,54 +407,96 @@ func (sdp *standardDelegationProcessor) verifyStakedValue(sc genesis.InitialSmar
...
-
-	vmOutputBlsKeys, err := sdp.queryService.ExecuteQuery(scQueryBlsKeys)
+	vmOutputStakeValue, err := sdp.queryService.ExecuteQuery(scQueryStakeValue)
 	if err != nil {
 		return err
 	}
+	if len(vmOutputStakeValue.ReturnData) != 1 {
+		return fmt.Errorf("%w return data should have contained one element", genesis.ErrWhileVerifyingDelegation)
+	}
+	scStakedValue := big.NewInt(0).SetBytes(vmOutputStakeValue.ReturnData[0])
```
Comment (sasurobert) [inline genesis/process/intermediate/standardDelegationProcessor.go:438]: add empty line before

File: genesis/process/testdata/smartcontracts.json [hunk 4]
```diff
@@ -10,8 +10,8 @@
     "owner": "0102030405060708090001020304050607080900010203040506070809000102",
     "filename": "testdata/delegation.wasm",
     "vm-type": "0500",
-    "init-parameters": "0BB8@%auction_sc_address%@0A61D0",
+    "init-parameters": "%auction_sc_address%@03E8@03E8@03E8",
```
Comment (sasurobert) [inline genesis/process/testdata/smartcontracts.json:13]: we need a tag somewhere and release version of contract before mainnet. something official.

File: integrationTests/multiShard/smartContract/scCallingSC_test.go [hunk 5]
```diff
@@ -519,11 +519,11 @@ func TestSCCallingInCrossShardDelegation(t *testing.T) {
 		t.Skip("this is not a short test")
 	}
 
-	numOfShards := 2
-	nodesPerShard := 3
-	numMetachainNodes := 3
-	shardConsensusGroupSize := 2
-	metaConsensusGroupSize := 2
+	numOfShards := 1
```
Comment (sasurobert) [inline integrationTests/multiShard/smartContract/scCallingSC_test.go:522]: move this values back.

File: genesis/process/testdata/smartcontracts.json (hunk 4 above)
Comment (iulianpascalau) [reply genesis/process/testdata/smartcontracts.json:13]: Totally agree, with hash verification. Let's plan this through.

File: genesis/process/intermediate/standardDelegationProcessor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply genesis/process/intermediate/standardDelegationProcessor.go:438]: Or create `delegationSCHandler` with the method `getUserStake()`, see comment below.

File: process/smartContract/process.go [hunk 6]
```diff
@@ -870,13 +871,15 @@ func (sc *scProcessor) createSCRForSender(
...
 	setOriginalTxHash(scTx, txHash, tx)
 
-	scTx.Data = []byte("@" + hex.EncodeToString([]byte(returnCode.String())))
-	for _, retData := range returnData {
-		scTx.Data = append(scTx.Data, []byte("@"+hex.EncodeToString(retData))...)
-	}
-
 	if callType == vmcommon.AsynchronousCall {
 		scTx.CallType = vmcommon.AsynchronousCallBack
+		scTx.Data = []byte("@" + hex.EncodeToString(big.NewInt(int64(returnCode)).Bytes()))
```
Comment (camilbancioiu) [inline process/smartContract/process.go:876]: This change might require a corresponding one in Arwen. Please let me know when you merge this.

File: genesis/process/intermediate/standardDelegationProcessor.go (hunk 1 above)
Comment (iulianpascalau) [reply genesis/process/intermediate/standardDelegationProcessor.go:169]: removed, remained from initial testing.

File: genesis/process/intermediate/standardDelegationProcessor.go (hunk 2 above)
Comment (iulianpascalau) [reply genesis/process/intermediate/standardDelegationProcessor.go:439]: deleted

File: genesis/process/intermediate/standardDelegationProcessor.go (hunk 3 above)
Comment (iulianpascalau) [reply genesis/process/intermediate/standardDelegationProcessor.go:438]: done

File: integrationTests/multiShard/smartContract/scCallingSC_test.go (hunk 5 above)
Comment (iulianpascalau) [reply integrationTests/multiShard/smartContract/scCallingSC_test.go:522]: Not my change, will revert.

Comment (iulianpascalau) [reply integrationTests/multiShard/smartContract/scCallingSC_test.go:522]: Reverted @sasurobert ' s changes :))))

File: genesis/process/intermediate/standardDelegationProcessor.go [hunk 7]
```diff
@@ -401,54 +407,96 @@ func (sdp *standardDelegationProcessor) verifyStakedValue(sc genesis.InitialSmar
...
-	return sdp.sameElements(vmOutputBlsKeys.ReturnData, nodesAddresses)
+	return nil
 }
 
-func (sdp *standardDelegationProcessor) sameElements(scReturned [][]byte, loaded [][]byte) error {
-	if len(scReturned) != len(loaded) {
-		return fmt.Errorf("%w staked nodes mismatch: %d found in SC, %d provided",
-			genesis.ErrWhileVerifyingDelegation, len(scReturned), len(loaded))
+func (sdp *standardDelegationProcessor) verifyOneNode(
+	sc genesis.InitialSmartContractHandler,
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply genesis/process/intermediate/standardDelegationProcessor.go:472]: I have stopped myself doing this as I have considered InitialAccountHandler and InitialNodesHandler, DelegationDataHandler, InitialSmartContractHandler as DTO's (that have interfaces over them). 0 functionality, just holding data. We can discuss further on this topic :coffee:

File: process/smartContract/process.go (hunk 6 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:876]: ok, will try to merge this asap.

File: genesis/process/intermediate/standardDelegationProcessor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply genesis/process/intermediate/standardDelegationProcessor.go:438]: will discuss about this, see below

File: genesis/process/intermediate/standardDelegationProcessor.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply genesis/process/intermediate/standardDelegationProcessor.go:454]: No need for implemebtation. It is overkill. DTOs are enough
